File: R/R/pmsc.R
#' Piecewise Multiplicative Scatter Correction (PMSC)
#'
#' @description
#' Perform piecewise multiplicative scatter correction (PMSC), i.e., MSC performed separately on a sliding band centered at the jth column of X
#'
#' @param X row vector or matrix of spectra to correct
#' @param w moving window size (must be an odd integer)
#' @param r reference spectrum
#' @param method Algorithm to calculate PSMC correction (see below for details)
#'
#' @return Matrix with PMSC-corrected spectra
#'
#' @details
#' 3 different algorithms for PMSC are supported using the 'methods' argument: "cpp" (default), "simple", and "mpmsc":
#'
#' "simple" will calculate the full MSC correction at a given moving window, and will extract the correction corresponding to the j-th wavelength.
#'
#' "cpp" calculates the MSC coefficients corresponding to the moving window surrounding the j-th column of X, and performs the correction using this manner.
#' This is all calculated inside a C++ wrapper function for speedup. This version is the fastest to calculate.
#'
#' "mpmsc" uses moving sums and moving means to calculate the correction; This version is optimal in MATLAB but is slightly slower in R
#'
#' @export
#' @examples
#' data("marzipan")
#' ref <- colMeans(marzipan$X)
#' X.pmsc <- mpmsc(marzipan$X,201,ref) #PMSC with window size of 201 channels
#' matplot(as.vector(marzipan$ax),t(X.pmsc),type="l",lty=1)
#' @references C. Giglio, J.-M. Roger, E. Andries. Piecewise Standard Normal Variate (PSNV) for scatter correction. J Chemometrics (2025) (Submitted)
#'
#' T. Isaksson, B. Kowalski, Applied Spectroscopy 1993, 47 (6), 702–709
#'
pmsc <- function( X, w, r, method=c("cpp","simple","mpmsc") ){
  if(!is.matrix(X)){ X <- as.matrix(X)}
  if(is.null(nrow(X))||ncol(X)==1){ #x is a vector or a column matrix (1 x n)
    X <- matrix(X,nrow=1) #convert to a row matrix
  }
  if(missing(r)){r <- colMeans(X)}
  n <- ncol(X)# = size(X,2);
  if(length(r)!=n){stop("ref must match number of columns of X")}
  if(w%%2 != 1){stop("w must be an odd positive integer")}
  if(w<=1){stop("w too small")}
  if(w>(2*ncol(X))){stop("w too large")}
  if(missing(method)){method <- "cpp"}
  if(method %in% c("cpp","simple","mpmsc")==F){stop("Argument 'method' should be one of “cpp”, “simple”, “mpmsc”")}
  r <- matrix(r,nrow=1) #Ensure r is a ROW vector
  #Pre-compute intermediate moving results:
  switch(match.arg(method), simple = {
    X.pmsc <- X
    for(j in 1:n){
      band <- max(1,j-v):min(j+v,n)
      Xmsc <- MSC(X[,band],r = r[band])
      X.pmsc[,j] <- Xmsc[,which(band==j)]
    }
  }, cpp = {
    X.pmsc <- pmsc_cpp(X,w,r)
  }, mpmsc = {
    X.pmsc <- mpmsc(X,w,r)
  })
  return(X.pmsc)
}

File: Matlab/pmsc.m
function Z = pmsc( X, v, r, classic )
%---------------------------------------------------------------------
% PURPOSE: Piecewise MSC, i.e., MSC performed on sliding window.
%---------------------------------------------------------------------
%
% INPUT:
% [1] X: (n,p) matrix of spectra.  Each spectrum in the matrix X 
%     is aligned row-wise.
%
% [2] v: integer bounded by 2 < v < p. 
%     Half-width of sliding window: max(1,j-v):min(j+v,p).
%
% [3] r: p-vector. (OPTIONAL) 
%     The reference spectrum. The default value for r is mean(X,1). 
%
% [4] classic:  logical. (OPTIONAL) 
%     If true, then do classical MSC, otherwise, do an
%     equivalent version formulated by Inge Helland, i.e.,  
%         Z = mean(r) + 1/b*(x-mean(x)).  
%     described in [1] in the mfile msc.m.  The default value is true. 
%
%---------------------------------------------------------------------
%
% OUTPUT:
% Z: Matrix of transformed spectra
%
%---------------------------------------------------------------------

%-------------------------------------------------
% DEFAULT INPUT ARGUMENTS
%-------------------------------------------------
p = size(X,2);
% Sanity check half-width
if (v<3) || (v-floor(v)~=0) || (v>(p-1))
    error('v must be integer bounded by 3 <= v <= size(X,2).');
end  
% Get reference spectrum 
if (nargin<3), r = []; end
if isempty(r), r = mean(X,1); end
r = r(:); % Force r to be column vector
p2 = length(r); 
if p2 ~= p
    error('length(r) does not equal size(X,2)'); 
end 
% Type of MSC implementation 
if (nargin<4), classic = []; end
if isempty(classic), classic = true; end
if ~islogical(classic)
    error('classic should be logical (true or false)')
end  

%-------------------------------------------------
% MSC ON SLIDING WINDOW
%-------------------------------------------------
Z = zeros(size(X));
for j = 1:p
    window = max(1,j-v):min(j+v,p);        
    Zj = msc( X(:,window), r(window), classic );   
    Z(:,j) = Zj(:,window==j);     
end

%---------------------------------------------------------------------
end